crypto/bn/bn_dsa.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+#include <openssl/opensslconf.h>
+#ifdef FIPS_MODE
+NON_EMPTY_TRANSLATION_UNIT
+#else
+# include <stdio.h>
+# include <string.h>
+# include <openssl/evp.h>
+# include <openssl/rand_drbg.h>
+# include "internal/cryptlib.h"
+# include "crypto/bn.h"
+# include "internal/thread_once.h"
+# include "../rand/rand_local.h"
+
+typedef struct rfc6979_seed_st
+{
+    const unsigned char *x;
+    const unsigned char *h;
+    size_t rlen;
+} RFC6979_SEED;
+
+static size_t rfc6979_get_entropy(RAND_DRBG *drbg, unsigned char **pout,
+                          int entropy, size_t min_len, size_t max_len,
+                          int prediction_resistance)
+{
+    RFC6979_SEED *rfc6979_seed = RAND_DRBG_get_callback_data(drbg);
+
+    *pout = (unsigned char *)rfc6979_seed->x;
+    return rfc6979_seed->rlen;
+}
+
+static size_t rfc6979_get_nonce(RAND_DRBG *drbg, unsigned char **pout,
+                        int entropy, size_t min_len, size_t max_len)
+{
+    RFC6979_SEED *rfc6979_seed = RAND_DRBG_get_callback_data(drbg);
+
+    *pout = (unsigned char *)rfc6979_seed->h;
+    return rfc6979_seed->rlen;
+}
+
+static int bits2int(BIGNUM *out, int qlen,
```

Comment:
Is BN_putbitsize a better name?  (And if so, does it move to the BN lib?)  Either way, needs an intro comment.

Reply:
The function names is used in https://tools.ietf.org/html/rfc6979#section-2.3
I thought these names are fine ,do you think that I really need to rename it?

Reply:
I am not a member of the project.  You should see what they say.

Reply:
Personally, I like the idea to refer to the original names used in the standard. A good compiler will inline the call anyway, so it does no harm. To make it more obvious, I would add a reference to the respective section, for example like this:

```C
/*
 * RFC 6979 2.3.2.  Bit String to Integer
 */
static int bits2int(BIGNUM *out, int qlen,
    const unsigned char *message, size_t message_len)
{
...
}
```

(@richsalz yes I know that I could have used a single line comment here. But somehow it looks better this way. ;-) )


We have a precedent for this kind of documentation; just look at this excellent example by @slontis:

https://github.com/openssl/openssl/blob/0e43960e88128bb86031a45c0fe9ca3e3a310c3b/crypto/rand/drbg_hmac.c#L60-L77


In this case of these little functions there is no need to add lengthy comments, a simple reference to the section suffices.

Reply:
I would also add a general comment with a reference to [RFC6979 Section 3.3.  Alternate Description of the Generation of k](https://tools.ietf.org/html/rfc6979#section-3.3) somewhere in the module.

Reply:
@mspncp OK, I will add it
But I started to doubt that whether name the file "crypto/bn/bn_dsa.c" is improper?
Should this file belong to BN module?

Reply:
No, it's a signature, so it doesn't belong in the bn directory.  But maybe dsa6979.c ?

Reply:
IMO rfc6979.c is better
But which directory should it belong to?
and the declaration of the function should be put into which header file?

Reply:
> No, it's a signature, so it doesn't belong in the bn directory. But maybe dsa6979.c ?

It's  _not_ a signature (algorithm), only a building block for it. And this building block is used for ECDSA and DSA, so the module fits neither in the ec nor the dsa subdirectory. I agree that the file is misnamed, but I'm not sure whether it is misplaced. Maybe @mattcaswell can make some suggestions how to name it and where  to put it?

Also the "dsa" should be dropped from`bn_generate_dsa_deterministic_nonce()`. Maybe rename it to `bn_generate_deterministic_nonce()` or `bn_generate_rfc6979_nonce()`?